### ISSUE TITLE
Partial document update based on JSON Patch

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#3166]
## Problem

Currently there's no way to do a partial document update. 

- https://github.com/PostgREST/postgrest/issues/465#issuecomment-1198884375 discusses json patch but there are certain operations (like `test`) which don't make much sense given our horizontal filtering. 
- An alternative is proposed on https://github.com/PostgREST/postgrest/issues/465#issuecomment-1199858002, but it's flexibility makes it harder to implement. Additionally it's not isolated to a single content type.

## Solution

Use a vendored media type based on json patch where we support the operations we lack. As prior art, [azure cosmosdb json patch](https://learn.microsoft.com/en-us/azure/cosmos-db/partial-document-update#supported-operations) (also see [here](https://learn.microsoft.com/en-us/rest/api/cosmos-db/patch-a-document)) does a similar thing and adds support for `increment` and `set`, which are not defined in the [JSON Patch RFC](https://www.rfc-editor.org/rfc/rfc6902.html).

For starters, we can add support for increment. Like so:

```http
PATCH /items?id=eq.1
Content-Type: application/vnd.pgrst.json-patch

[
  { "op": "incr", "path": "/inventory", "value": 10 },
]
```

Path would refer to any column and the generated SQL would be like `UPDATE items SET inventory = inventory + 10`.

Later on we can extend this to add other operations, like `mult`, `div`, `concat`, etc.

**Comment by JanKaul** [comment]
Am I assuming correctly that the implementation would involve a new [media type handler](https://postgrest.org/en/stable/references/api/media_type_handlers.html#media-type-handlers) and a [postgres implementation of jsonpatch](https://gist.github.com/InfoSec812/b830a9db4c9048552f8c51d7987cc4d0)?

What other steps would be required to implement this functionality as a builtin media type?

**Comment by JanKaul** [comment]
That makes sense. Thanks for the clarification. I will have a look at the implementation of `application/vnd.pgrst.plan`.

**Comment by steve-chavez** [comment]
Taking advantage of the [interval type](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT) in postgres, it would be interesting to have an `interval_pl` operator :

```http
PATCH /items?id=eq.1
Content-Type: application/vnd.pgrst.json-patch

[
  { "op": "interval_pl", "path": "/date", "value": "1 day 15 minute 40 seconds"},
]

``` 

That would get translated roughly to:

```sql
UPDATE items SET date = date + '1 day 15 minute 40 seconds'::interval;
```

Interval already handles negatives (can set `-'1 day'::interval`), so no need for an additional operator for substractions.

The `interval_pl` name was taken from the `pg_operator` table.

```sql
select * from pg_operator where oprcode = 'interval_pl'::regproc;
```

Maybe we can name it just `intervalpl` or `interval_incr`/`intervalincr`.

**Comment by steve-chavez** [comment]
Since we're only based of JSON patch, we don't really need to use `path` as defined by them. So we could just use the field without the `/` prefix:

```json
[
  { "op": "interval_pl", "path": "date", "value": "1 day 15 minute 40 seconds"},
]
```

That would save us parsing and let us adjust the `path` to have meaning according to the operator, for example later on for `jsonb_set`:

```json
[
  { "op": "set", "path": "col->x->y", "value": 10 },
]
```

**Comment by taimoorzaeem** [comment]
I am not sure if we should strictly follow [RFC 6902](https://datatracker.ietf.org/doc/html/rfc6902) here. According to RFC, it should also support multiple patches in a single request as described [here](https://datatracker.ietf.org/doc/html/rfc6902/#section-5):

```json
[
   { "op": "replace", "path": "/a/b/c", "value": 42 },
   { "op": "test", "path": "/a/b/c", "value": "C" }
]
```
We always try to be standard compliant but I guess we are slightly off track here. To what extent are we following the RFC here?

**Comment by steve-chavez** [comment]
```json
[
   { "op": "replace", "path": "/a/b/c", "value": 42 },
   { "op": "test", "path": "/a/b/c", "value": "C" }
]
```

We should allow multiple operations. The `test` operation is not needed as that is replaced by our filters.

Check more comments on https://github.com/PostgREST/postgrest/issues/465#issuecomment-1198884375

**Comment by taimoorzaeem** [comment]
I'd say it would be best to implement this feature as cleanly as possible since we already have quite a lot tech debt because of pending refactors.

For starters, how about adding a new patch `MutatePlan` here:

```haskell
data MutatePlan
  = Insert { ...
    }
  | ...
  | JsonPatch {  -- new json patch mutate plan
      ...
   }
```
Also, we can have a parsed payload here because jsonpatch has a fixed json format:

```haskell
data Payload
  = ProcessedJSON -- ^ Cached attributes of a JSON payload
      { payRaw  :: LBS.ByteString
      , payKeys :: S.Set Text
      }
  | ProcessedUrlEncoded { payArray  :: [(Text, Text)], payKeys :: S.Set Text }
  | RawJSON { payRaw  :: LBS.ByteString }
  | RawPay  { payRaw  :: LBS.ByteString }
  | JsonPatchPayload { payJsonPatch :: ParsedJsonPatchPayload } -- something like that

```

Not sure about this but I guess maybe we can give this a try and see?

**Comment by steve-chavez** [comment]
> For starters, how about adding a new patch MutatePlan here:
> Also, we can have a parsed payload here because jsonpatch has a fixed json format:

Yes, that looks good.

> since we already have quite a lot tech debt because of pending refactors.

I think the Plan and QueryBuilder are mostly devoid of tech debt, but they do require a lot of work, the logic is always complex there.

**Comment by steve-chavez** [comment]
I've noticed that this interface has a drawback in observability, so our current apache logs would show as 

```
127.0.0.1 - postgrest_test_anonymous [03/Apr/2026:18:21:07 -0500] "PATCH /projects HTTP/1.1" 200 212 "" "curl/7.81.0"
```

i.e. the exact operation can't be known because it's hidden inside the json payload. And if we wanted to show this on the logs in another format, we would have to parse the json body resulting in perf loss.

Also users can send really large json objects and us having to parse inside the json to know the operation will kill performance.

The previous proposal with the URL syntax  https://github.com/PostgREST/postgrest/issues/465#issuecomment-1199858002 doesn't have this drawback.

**Comment by wolfgangwalther** [comment]
> I've noticed that this interface has a drawback in observability, so our current apache logs would show as [...] i.e. the exact operation can't be known because it's hidden inside the json payload.

I don't think we should consider this a blocker for the feature in general. It will *still* be a PATCH, that is not incorrect.

**Comment by steve-chavez** [comment]
> I don't think we should consider this a blocker for the feature in general.

The other part I added:

> Also users can send really large json objects and us having to parse inside the json to know the operation will kill performance.

Is IMO a much bigger reason to not do it, we already suffer from performance issues on big payloads (considering double copies and all). With the other URL syntax way we reduce tremendously the amoung of parsing we have to do for this type of requests.
